### ISSUE TITLE
How to detect an connection failure?

### ISSUE THREAD

**Issue by twity1337** [Marfusios/websocket-client#27]
Hello,
could you please add more error handling examples to the documentation? For instance, how can I detect if the server is not available. There should be some kind of Exception or something when calling start(), saying that connection attempt wasn't successful.

Instead it seems, that nothing happens.

Thanks in advance.

**Comment by Marfusios** [comment]
Hello, 

good question, I will update documentation. 
But in short description, there is `DisconnectionHappened` stream, which publishes a message on every disconnection and every unsuccessful try (with type: `DisconnectionType.Error`). 

In case of a very first connection, you can await `Start()` method, which blocks execution until it connects to the server. But it is quite uncomfortable in a situation of long server outage, so it is better to call `Start()` without await/Wait() and handle messages inside streams `DisconnectionHappened` and `ReconnectionHappened`. 

```csharp
var client = new WebsocketClient(url, factory);
client.ErrorReconnectTimeoutMs = (int)TimeSpan.FromSeconds(30).TotalMilliseconds;
client.DisconnectionHappened.Subscribe(type => 
    Log.Warning($"Disconnection happened, type: {type}") // called every 30 sec, with type 'Error'
    );

Log.Information("Starting...");
await client.Start();  // blocks execution
Log.Information("Started.");
```

Also, there is property `IsRunning` which returns true only if websocket client is connected to the server.

**Comment by twity1337** [comment]
@Marfusios Wouldn't it be better, if an exception is thrown by `Start()` if a connection couldn't be established on the first try? What do you think about that approach?

**Comment by Marfusios** [comment]
Understand your point of view, probably the best solution would be to provide a second optional method, something like `await StartOrFail()`. 
Because in all my use-cases I don't wanna bother with exceptions/errors when starting, but I see that it could be handy for some situation when you want to react accordingly. 

What do you think?

**Comment by dthisner** [comment]
is it a way to get back what the error message is? 
Right not I am getting an exception saying: "Unable to connect to the remote server", would be nice to have something more then just `Error` :)

**Comment by Marfusios** [comment]
@twity1337 @dthisner implemented in [version 4.0](https://github.com/Marfusios/websocket-client/releases/tag/v4.0)
